Attestation Updates (#17)
riscv-non-isa/riscv-ap-tee: specification/attestation.adoc

Relying Parties in their decision process.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Remote Attestation Framework (IETF RATS)", pdfwidth=75%]
image::rats-arch.svg[]

==== Layered Evidence

The AP-TEE Attestation Evidence is a collection of Claims representing the
overall workload TCB and state, which includes those of the platform, TSM
and TVM. It uses the DICE<<DICE>> layered attestation model where each TCB layer
collects Claims about the layers it loads, acting as the Attestating Environment
for them. As the root TCB layer, the hardware RoT is the initial Attestating
Environment and its own Claims are asserted by an Endorser.

In the DICE model, each TCB layer computes the Compound Device Identifier (CDI)
for the layer it loads. CDIs are secrets that must not leave the AP-TEE TCB. The
root TCB layer, i.e. the hardware RoT, is factory provisioned with a Unique
Device Secret (UDS) that serves as the original secret from which all other CDIs
are directly or indirectly derived. As such, the TVM CDI is rooted into the
provisioned and platform manufacturer endorsed UDS.
The CDI for the current TCB layer is a cryptographic derivation of a combination

> of the Attestating Environment CDI (i.e. the loading TCB layer) and the current

Author: Fixed as well, thanks.

The Platform EAT is a CWT tagged CBOR formatted token, wrapped with a
COSE_Sign1 <<COSE>> envelope. It is signed by the RoT attestation key and
must contain a hash of the TSM public key, in order to create a cryptographic
bound between the TSM and the platform tokens.

The AP-TEE layered Evidence structure is represented by the above described
composition of cryptographically chained EAT tokens. Verifier can then attest
of an AP-TEE workload trustworthiness by independenly inspecting each token,
while being able to verify that the TCB represented by one token was used to
generate the next one.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "AP-TEE Attestation Evidence Format", pdfwidth=65%]
image::attestation_token.svg[]

The AP-TEE Attestation Evidence is defined by the following Concise Data
Definition Language (CDDL):

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "AP-TEE Attestation Evidence CDDL"]
[source,cddl]
----
COSE_Sign1 = #6.18(COSE_Sign1_Payload)
submodule-label = 266 ; EAT Submodule

; Each token is a nested CWT
platform-token = COSE_Sign1
tsm-token = COSE_Sign1
tvm-token = COSE_Sign1

; One EAT Submodule

> riscv-ap-tee-token = (
>    submodule-label => {
>       "platform" => platform-token
>       "tsm" => tsm-token
>       "tvm" => tvm-token
>    }
> )

Reviewer: I think you want a map here, not a group, right?

```suggestion
riscv-ap-tee-token = {
   submodule-label => {
      "platform" => platform-token
      "tsm" => tsm-token
      "tvm" => tvm-token
   }
}
```

Author: Yep, this is a map now.

                       bytes .size 48 / ; SHA384
                       bytes .size 64   ; SHA512

platform-challenge = (
   platform-challenge-label => riscv-ap-tee-hash-type
)
----

====== Platform Public Key Claim

The platform public key claim identifies the platform public key that signs the
TSM token. The value of the claim is a COSE_Key <<COSE>> structure encoded as a
CBOR bytes string.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Platform Public Key Claim"]
[source,cddl]
----
platform-public-key-label = TBD
riscv-ap-tee-public-key-type = bytes .cbor COSE_Key

platform-public-key = (
   platform-public-key-label => riscv-ap-tee-public-key-type
)
----

====== Platform Manufacturer Identifier Claim

The platform manufacturer identifier claim uniquely identifies the manufacturer
for the AP-TEE implementation providing the Attestation Evidence.
Verification services can use the value of this claim to obtain the manufacturer
provided Endorsements for the platform implementation.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Platform Manufacturer Identifier Claim"]
[source,cddl]
----
platform-manufacturer-id-label = TBD

> platform-manufacturer-id-type = .bytes 64

Author: Fixed, thanks.

> platform-manufacturer-id = {
>    platform-manufacturer-id-label => platform-manufacturer-id-type
> }

Reviewer: here instead you probably want a group rather than a map

```suggestion
platform-manufacturer-id = (
   platform-manufacturer-id-label => platform-manufacturer-id-type
)
```

Author: And that is a group now, thanks.

----

====== Platform State Claim

The platform state claim describes the operational state of the platform.
The values for this claim can be:

* `NotConfigured` - The platform requires additional information to operate
* `Secured` - This is the default state under regular operation conditions
* `Debug` - The platform can be debugged
* `Recovery` - The platform is recovering from a failure

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Platform State Claim"]
[source,cddl]
----
platform-state-label = TBD
platform-state-not-configured = 1
platform-state-secured = 2
platform-state-debug = 3
platform-state-recovery = 4
platform-state-type =
                    platform-state-not-configured \
                    platform-state-secured \
                    platform-state-debug \

>                     platform-state-recovery

Author: Ah yes, good catch. Fixed now.

component.
* *Hash algorithm identifier*: This describes which algorithm was used to generate the
component measurement. It is recommended to use one of the IANA defined
<<HashAlgorithmNames>>.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Platform Software Components Claim"]
[source,cddl]
----
platform-sw-components-label = TBD
riscv-ap-tee-sw-component = {
   1 => text                   ; Component type
   2 => riscv-ap-tee-hash-type ; Component measurement value
   3 => text                   ; Component Secure Version Number (SVN)
 ? 4 => riscv-ap-tee-hash-type ; Component manifest hash
   5 => riscv-ap-tee-hash-type ; Component signer public key hash value
   6 => text                   ; Hash algorithm identifier
}

platform-sw-components = (
   platform-sw-components-label => [ + riscv-ap-tee-sw-component ]
)
----

===== TSM Token

The TSM Token is a nested EAT token in the Evidence and contains a map of
Claims. The map is wrapped in a `COSE_Sign1` envelope and composed of the
following claims:

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TSM Claims Map"]
[source,cddl]
----

> tsm-token = #6.61(tsm-token-map)

Author: I was not sure how to describe a `COSE_Sign1`, but thanks to your [suggestion](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/tree/main/cddl) I now have a better idea. So this now looks like that:

```cddl
submodule-label = 266 ; EAT Submodule

protected-ap-tee-token-header-map = {
  alg => int    ; See https://www.iana.org/assignments/cose/cose.xhtml
  ? kid => bstr ; Optional, only needed for the platform token
}

unprotected-ap-tee-token-header-map = {
  * int / tstr => any
}

signed-ap-tee-token = #6.18(COSE-Sign1-ap-tee-token)
COSE-Sign1-ap-tee-token = [
  protected: bstr .cbor protected-ap-tee-token-header-map
  unprotected: unprotected-ap-tee-token-header-map
  payload: bstr .cbor platform-token / tsm-token / tvm-token
  signature: bstr
]

; One EAT Submodule
; Each token is a nested, signed CWT
riscv-ap-tee-token = {
   submodule-label => {
      "platform" => signed-ap-tee-token
      "tsm" => signed-ap-tee-token
      "tvm" => signed-ap-tee-token
   }
}
```

an argument to the `sbi_tee_guest_get_evidence()` SBI call.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TSM Challenge Claim"]
[source,cddl]
----
tsm-challenge-label = 10 ; EAT nonce

tsm-challenge = (
   tsm-challenge-label => riscv-ap-tee-hash-type
)
----

====== TSM Public Key Claim

The TSM public key claim identifies the platform public key that signs the
TVM token. The value of the claim is a COSE_Key <<COSE>> structure encoded as a
CBOR bytes string.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TSM Public Key Claim"]
[source,cddl]
----
tsm-public-key-label = TBD

tsm-public-key = (
   tsm-public-key-label => riscv-ap-tee-public-key-type
)
----

====== TSM Software Components Claim

The TSM software components claim lists all software components that compose the
AP-TEE TSM TCB.

The TSM software components that influence the TSM TCB are the TSM-Driver and the
TSM.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TSM Software Components Claim"]
[source,cddl]
----
tsm-sw-components-label = TBD

tsm-sw-components-type = [
   tsm-driver
   tsm
]

tsm-driver = (riscv-ap-tee-sw-components)

> tsm = (riscv-ap-tee-sw-components)

Reviewer: It looks like `riscv-ap-tee-sw-components` is not defined

Author: Right, I meant `riscv-ap-tee-sw-component`.  Should be fixed now.

tsm-sw-components = (

>    tsm-sw-components-label => tsm-sw-component-type

Author: Fixed, thanks.

:imagesdir: ./images

[[attestation]]
== TVM Attestation and Measurements

The CoVE TVM attestation framework allows for CoVE workload owners to assert
the trustworthiness of the hardware and software environment their workload is
running in.

Attestation relies on the ability for the SoC to generate a cryptographic evidence
for a workload executing in a CoVE TVM. The workload executing in a TVM may
request this cryptographic evidence to relay to a remote relying party which can
then verify that the evidence is valid (per some appraisal policy), and thus attest
to the trustworthiness of the TVM. The relying party can then accept to release
secrets or attestation result tokens back to the trusted workload.

This section describes the CoVE attestation evidence content, format and
generation interface.

=== Measurements

A CoVE workload measurement is a cryptographic hash of at least one of the
TCB elements for the workload. Measurements can only be extended with additional
data, by applying a cryptographic hash algorithm (`H~alg~`) to the current
measurement value (`M~current~`):

:stem: asciimath
[.center]
[stem]
++++
M_{extended} = H_{alg}(M_{current} || NewData)
++++

> Measurements must be stored in integrity protected measurement registers,

Reviewer 2: ```suggestion
Measurements must be stored in integrity protected measurement registers, or in confidential memory,
```

> either in the Hardware RoT of Measurement (RoTM) or in the TSM when it acts as

Reviewer 2: ```suggestion
either in the Hardware RoT of Measurement (RoTM) or in the TSM (in confidential memory) when the TSM acts as
```

an extended RoTM.

In order to verify a workload environment trustworthiness, the remote
attestation service must be able to compare the CoVE TCB measurements with a
set of reference values. All components of what constitues the workload Trusted
Computing Base (TCB) must be measured and included in the attestation evidence.

==== TCB Elements

Elements considered to be in the TCB for CoVE workloads are both hardware
and software components.

The overall workload TCB is a composition of 3 independent TCBs:

1. The Platform TCB is made of all hardware and software TCB elements for
the CoVE host platform. They are not confidential compute specific and
measurements for those elements could be shared with e.g. a measured boot
attestation.

> 2. The TSM TCB elements are the confidential compute specific but TVM agnostic

Reviewer 2: ```suggestion
2. The TSM TCB elements are confidential-computing specific but TVM agnostic
```

elements for the overall TCB.
3. The TVM TCB is composed of the TVM specific elements.

The TCB elements for each of them is summarized in the following table:

.COVE Workload TCB Elements
[cols="^1, ^5v, ^4v", options="header"]
|===
|TCB |Hardware Elements |Software Elements

.4+.^|Platform
    <| HW RoT for boot, measurement and storage
 .4+<| All M-mode firmwares, including the TSM-driver
    <| All CPU hardware logic, including MMU and caches

>     <| All SoC subsystems, including memory confidentiality, integrity and replay-protection for volatile memory

Reviewer 2: ```suggestion
    <| All SoC subsystems, including subsystems that may enforce memory confidentiality, integrity and replay-protection for volatile memory
```

    <| IOMMU and translation agents

     | TSM
     | N/A
     | TEE Security Manager (TSM) and its U-mode components

     | TVM
     | Directly assigned, TEE-IO compliant devices
     | All TVM measured pages
|===

==== TVM Measurements

At a high level, a TVM measurement is separated into a set of static

> measurements and a dynamic one.

Reviewer 2: ```suggestion
measurements and a set of dynamic measurements.
```

===== Static Measurements

TVM static measurements are generated from the CoVE workload TCB elements

> involved in the TVM construction. Any TCB element that directly or indirectly

Reviewer 2: ```suggestion
involved in the TVM build process. Any TCB element that directly or indirectly
```

supports a TVM must be measured into the TVM static measurement registers. Once

> a TVM is finalized, i.e. after the `sbi_tee_host_finalize_tvm()` TH-ABI is

Reviewer 2: ```suggestion
a TVM build is finalized, i.e. after the `sbi_tee_host_finalize_tvm()` TH-ABI is
```

The static TCB measurement mechanism and storage layout is implementation
specific. One such possible layout, based on a 6 static measurement registers
CoVE implementation may look like the following table:

.Static Measurements Layout Example (6 registers)
[cols="^1,^3,^5,^1", options="header"]
|===
|Register |Usage |TCB Elements |Storage
.4+.^| 0
.4+.^| Platform Firmware
    <| RoT Firmware
.4+.^| RoT
    <| All SoC Subsystems Firmwares
    <| All Regular M-mode Firmwares
    <| TSM-Driver

     | 1
     | Platform Configuration
    <| All Firmware Manifests
     | RoT

     | 2
     | TSM Software
    <| TSM Binary
     | RoT

     | 3
     | TSM Configuration
    <| TSM Manifest
     | RoT

     | 4
     | TVM Code and Static Data
    <| TVM Measured Pages
     | TSM

     | 5
     | TVM Configuration
    <| TVM Entry Point and Initial Arguments
     | TSM
|===

===== Dynamic Measurements

After the TVM is finalized, static measurements are immutable and can no longer
be modified. However, a TVM guest may want to measure some of its software
components (Guest firmware, kernel OS, additional configuration parameters, etc)
while it is booting or running.

> The TSM optionally provides an interface for TVM guests to extend their

Reviewer 2: ```suggestion
The TSM may provides an interface for TVM guests to extend their
```

(Evidence) to enable a remote peer (the Relying Party) to decide whether to
consider that Attester a trustworthy peer or not.
The Verifier authenticates the Evidence with Endorser provided trust anchors
(Endorsements), compares it against Reference Values and appraises it via
appraisal policies. It eventually creates Attestation Results to support
Relying Parties in their decision process.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title="Remote Attestation Framework (IETF RATS)", pdfwidth=75%]
image::rats-arch.svg[]

==== Layered Evidence

The CoVE Attestation Evidence is a collection of Claims representing the
overall workload TCB and state, which includes those of the platform, TSM
and TVM. It uses the DICE<<DICE>> layered attestation model where each TCB layer
collects Claims about the layers it loads, acting as the Attesting Environment
for them. As the root TCB layer, the hardware RoT is the initial CoVE Attesting
Environment and its own Claims are asserted by an Endorser.

In the DICE model, each TCB layer computes the Compound Device Identifier (CDI)
for the layer it loads. CDIs are secrets that must not leave the CoVE TCB. The

> root TCB layer, i.e. the hardware RoT, is factory provisioned with a Unique

Reviewer 2: ```suggestion
root TCB layer, i.e. the hardware RoT, is factory-provisioned with a Unique
```

Device Secret (UDS) that serves as the original secret from which all other CDIs
are directly or indirectly derived. As such, the TVM CDI is rooted into the
provisioned and platform manufacturer endorsed UDS.
The CDI for the current TCB layer is a cryptographic derivation of a combination
of the Attesting Environment CDI (i.e. the loading TCB layer) and the current
TCB layer measurements:

:stem: asciimath
[.center]
[stem]
++++
CDI_{0} = KDF(UDS_{Len},\ UDS\ ||\ H_{alg}(Meas(TCB_{0}))
++++
:stem: asciimath
[.center]
[stem]
++++
CDI_{N} = KDF(CDI_{Len},\ CDI_{N-1}\ ||\ H_{alg}(Meas(TCB_{N}))
++++

Asymmetric key pairs can be derived from a CDI in order to generate the

> attestation key for a DICE layer. This key can then be used to attest of the
> next TCB layer.

Reviewer 2: ```suggestion
attestation key for a DICE layer. This key can then be used to attest to the
TCB of the DICE layer.
```

.AP-TEE Workload TCB Elements
[cols="^1, ^5v, ^4v", options="header"]
|===
|TCB |Hardware Elements |Software Elements

.4+.^|Platform
    <| HW RoT for boot, measurement and storage
 .4+<| All M-mode firmwares, including the TSM-driver
    <| All CPU hardware logic, including MMU and caches
    <| All SoC subsystems, including memory confidentiality, integrity and replay-protection for volatile memory
    <| IOMMU and translation agents

     | TSM
     | N/A
     | TEE Security Manager (TSM) and its U-mode components

     | TVM
     | Directly assigned, TEE-IO compliant devices
     | All TVM measured pages
|===

==== TVM Measurements

At a high level, a TVM measurement is separated into a set of static
measurements and a dynamic one.

===== Static Measurements

TVM static measurements are generated from the AP-TEE workload TCB elements
involved in the TVM construction. Any TCB element that directly or indirectly
supports a TVM must be measured into the TVM static measurement registers. Once
a TVM is finalized, i.e. after the `sbi_tee_host_finalize_tvm()` TH-ABI is
called, the TVM static measurements must no longer be extended.

> Static measurements are stored in dedicated measurement registers and an AP-TEE

Reviewer 2: ```suggestion
Each TVM's static measurements are stored in dedicated measurement registers and an AP-TEE
```

implementation must provide at least 1 and at most 8 of them. The static
measurement registers cover platform, TSM & TVM specific TCB element

> measurements.

Reviewer 2: is this minimum 1 to max 8 for each TVM? then should qualify it as "TVM static measurements"

Author: Thanks, fixed.

remote attestation only;
  ** If OSAM is a 3rd party - the certifying entity will need a separate 
evidence entry. 
* Cryptographic hash of the TVM static binaries and configuration, along 
with its SVN and other fields.
* The TVM may additionally extend cryptographic measurements for other 
workload binaries and configuration loaded dynamically subsequent to boot 
via the TG-ABI.

The TVM TCB Evidence Info is managed by the TSM and is combined with the 
TSM’s TCB Evidence info that is in turn managed by the TSM-driver. The 
TSM-driver provides a TEEI security routine to enable the TSM and 
transitively the TVM to generate an Attestation CDI (Composite Device 
Identifier) and key to participate in an Attestation certificate-based 
protocol for remote (and local) attestation.

We recommend at least the following CDIs to be supported for AP-TEE 
workloads:

. Attestation CDI - This CDI is derived from the combination of the input 
values listed above and is expected to change across software updates or 
configuration changes of these components. This CDI is meant for remote 
attestation and is mandatory for AP-TEE implementations.

> . Versioned Sealing CDI - This CDI is also derived from the combination of 

Reviewer 2: There is no specification of versioned sealing CDI in this new attestation specification.

Author: Good point, I will add it as a separate commit.

Reviewer 2: ok

>      | TVM Code

Reviewer 2: ```suggestion
     | TVM Code and Static Data
```

Author: Fixed, thanks.

> components (Guest firmware, kernel OS) while it is booting.

Reviewer 2: ```suggestion
components (Guest firmware, kernel OS, additional configuration parameters) while it is booting.
```

Author: Fixed, thanks.

When loading the next TCB layer, the attesting layer provides it with both its
CDI and its attestation certificate.

The TVM guest can obtain an Attestation Evidence from the TSM by calling the
TG-ABI `sbi_tee_guest_get_evidence()` FID. The guest provides a challenge value
to the TSM through this call, and that value must be included in the generated
Evidence. This value allows relying parties to establish the Attestation
Evidence freshness.

The Evidence is signed by the TSM attestation key, which is HW RoT-rooted.
A Relying Party can then verify and authenticate the Evidence with platform
owner or manufacturer provided Endorsements, like e.g. a Trust Anchor.

The TCB extension and evidence collection for a TVM attestation is shown
below:

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Layered Attestation architecture for TVMs"]
image::layered_attestation.svg[]

It is expected that an implementation will provide implementation-specific
intrinsics to record measurements of the TSM into the firmware RoT for
measurement to support the AP-TEE layered RTMs attestation of AP-TEE workloads.

==== Evidence Format

> The AP-TEE Attestation Evidence is a profiled IETF Entity Attestation Token
> (<<EAT>>), formatted as an untagged Concise Binary Object Representation
> (<<CBOR>>) Web Token (<<CWT>>).

Reviewer 2: ```suggestion
The AP-TEE Attestation Evidence uses the IETF Entity Attestation Token (<<EAT>>), formatted as an untagged Concise Binary Object Representation (<<CBOR>>) Web Token (<<CWT>>).  An AP-TEE EAT profile is proposed to narrow the EAT specification for the AP-TEE use case to enable interoperability.
```

Author: Fixed, thanks.

The CWT is composed of one EAT submodule Claims-Set map where the map values
are nested tokens for the TVM, TSM and Platform Claims.

The TVM EAT is a CWT tagged CBOR formatted token, wrapped with a
COSE_Sign1 <<COSE>> envelope. It is signed by the TSM attestation key and must
contain a TVM provided challenge, e.g. a Relying Party provided nonce for
establishing Evidence freshness.

The TSM EAT is a CWT tagged CBOR formatted token, wrapped with a
COSE_Sign1 <<COSE>> envelope. It is signed by the Platform attestation key and
must contain a hash of the TVM public key, in order to create a cryptographic

> bound between the TVM and the TSM tokens.

Reviewer 2: ```suggestion
binding between the TVM and the TSM tokens.
```

The Platform EAT is a CWT tagged CBOR formatted token, wrapped with a
COSE_Sign1 <<COSE>> envelope. It is signed by the RoT attestation key and
must contain a hash of the TSM public key, in order to create a cryptographic

> bound between the TSM and the platform tokens.

Reviewer 2: ```suggestion
binding between the TSM and the platform tokens.
```

Author: Fixed, thanks.

[source,cddl]
----
platform-state-label = TBD
platform-state-not-configured = 1
platform-state-secured = 2
platform-state-debug = 3
platform-state-recovery = 4
platform-state-type =
                    platform-state-not-configured \
                    platform-state-secured \
                    platform-state-debug \
                    platform-state-recovery

platform-state = (
   platform-state-label => platform-state-type
)
----

====== Platform Software Components Claim

The platform software components claim lists all software and firmware
components that compose the AP-TEE platform TCB.

Each component in the Claim is a map of the following values:

* *Component type*: This is a human-readable string that represents the measured
component.
* *Component measurement*: This is the hash value of the component as it was
loaded in memory.
* *Component SVN*: This is the component Secure Version Number. The semantics of
that value is defined by the component creator or vendor. The SVN can be
generated from multiple SVNs.
* *Component manifest*: This is the hash value of the component manifest that was
used when loading and verifying the associated component. This field is

> _optional_.

Reviewer 2: ```suggestion
_optional_. The format of the manifest for the component is out of scope of this specification.
```

Author: Good point, thanks.

CBOR bytes string.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TSM Public Key Claim"]
[source,cddl]
----
tsm-public-key-label = TBD

tsm-public-key = (
   tsm-public-key-label => riscv-ap-tee-public-key-type
)
----

====== TSM Software Components Claim

The TSM software components claim lists all software components that compose the
AP-TEE TSM TCB.

The TSM software components that influence the TSM TCB are the TSM-Driver and the
TSM.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TSM Software Components Claim"]
[source,cddl]
----
tsm-sw-components-label = TBD

tsm-sw-components-type = [
   tsm-driver
   tsm
]

tsm-driver = (riscv-ap-tee-sw-components)
tsm = (riscv-ap-tee-sw-components)

tsm-sw-components = (
   tsm-sw-components-label => tsm-sw-component-type
)
----

===== TVM Token

The TVM Token is a nested EAT token in the Evidence and contains a map of
Claims. The map is wrapped in a `COSE_Sign1` envelope and composed of the
following claims:

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TVM Claims Map"]
[source,cddl]
----
tvm-token = #6.61(tvm-token-map)

> tsm-token-map = {

Reviewer 2: ```suggestion
tvm-token-map = {
```

Author: Ah, copy and paste error...good catch.

as a CBOR bytes string.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TVM Public Key Claim"]
[source,cddl]
----
tvm-public-key-label = TBD

tvm-public-key = (
   tvm-public-key-label => riscv-ap-tee-public-key-type
)
----

====== TVM Static Measurements Claim

The TVM static measurements claim value is the list of all static measurements
for the TVM. The list must contain at most 8 entries.

Each measurement in the list is a map of the following values:

* *Measurement register index*: This describes the measurement register index
used by the TSM to store the measurement value. This can be mapped to well known
measurement register indexes like e.g. the TCG<<TCG_Client>> defined ones.
* *Measurement value*: This is the measurement value.
* *Hash algorithm identifier*: This describes which algorithm was used to
generate the component measurement. It is recommended to use one of the IANA
registered hash algorithm name<<Hash_Algorithm_Names>>.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "TVM Static Measurements Claim"]
[source,cddl]
----
tvm-static-measurements-label = TBD
riscv-ap-tee-measurement = {

>  ? 1 => uint                   ; Measurement register index

Reviewer 2: is the index optional? I did not think so - then remove the "?"

Author: That's right, there will always be an index.

:imagesdir: ./images

:stylesheet: inline
[subs="attributes"]
++++
<style>
.center {
  text-align: center;
}
</style>
++++

[[attestation]]
== TVM Attestation and Measurements

The AP-TEE TVM attestation framework allows for AP-TEE workload owners to assert
the trustworthiness of the hardware and software environment their workload is
running in.

> Attestation relies on the ability for the SoC to generate a cryptographic
> evidence for an AP-TEE workload environment. A remote and workload owner trusted
> service can then verify that evidence and attest to the workload trustworthiness.
> This relying party can then accept to release secrets or attestion result tokens
> back to the trusted workload.

Reviewer 2: ```suggestion
Attestation relies on the ability for the SoC to generate a cryptographic evidence for a workload executing in an AP-TEE TVM. The workload executing in a TVM may request this cryptographic evidence to relay to a remote relying party which can then verify that the evidence is valid (per some appraisal policy), and thus attest to the trustworthiness of the TVM. The relying party can then accept to release secrets or attestation result tokens back to the trusted workload.
```

Author: Applied, thanks.

attestation service must be able to compare the AP-TEE TCB measurements with a
set of reference values. All components of what constitues the workload Trusted
Computing Base (TCB) must be measured and included in the attestation evidence.

==== TCB Elements

Elements considered to be in the TCB for AP-TEE workloads are both hardware
and software components.

The overall workload TCB is a composition of 3 independent TCBs:

1. The Platform TCB is made of all hardware and software TCB elements for
the AP-TEE host platform. They are not confidential compute specific and
measurements for those elements could be shared with e.g. a measured boot
attestation.
2. The TSM TCB elements are the confidential compute specific but TVM agnostic
elements for the overall TCB.
3. The TVM TCB is composed of the TVM specific elements.

The TCB elements for each of them is summarized in the following table:

|===
|TCB |Hardware Elements |Software Elements
|One 2+|Two and three |A \| separator character
|===

.AP-TEE Workload TCB Elements
[cols="^1, ^5v, ^4v", options="header"]
|===
|TCB |Hardware Elements |Software Elements

.4+.^|Platform
    <| HW RoT for boot, measurement and storage

>  .4+<| All M-mode firmwares, except for the TSM-driver

Reviewer 2: do you mean - including the TSM-driver?

Author: I mean to exclude the TSM-driver from the platform's TCB. I look at the platform TCB as containing all TCB elements that are not AP-TEE specific, which excludes the TSM-Driver.

Now, thinking a bit more about it: In the context of this specification (AP-TEE), the platform *will* have a `TSM-driver`, so I guess we should add it to the platform TCB and not the TSM one as even non TSM/TVM elements could be influenced by this M-mode component. I fixed it.

    <| All CPU hardware logic, including MMU and caches
    <| All SoC subsystems, including memory confidentiality, integrity and replay-protection for volatile memory
    <| IOMMU and translation agents

.2+.^| TSM
.2+.^| N/A
    <| TEE Security Manager (TSM) and its U-mode components
    <| TSM-Driver that hosts a TEEI (with TH-ABI and TG-ABI security routines).

     | TVM
     | Directly assigned, TEE-IO compliant devices
     | All TVM measured pages
|===

==== TVM Measurements

At a high level, a TVM measurement is separated into a set of static
measurements and a dynamic one.

===== Static Measurements

TVM static measurements are generated from the AP-TEE workload TCB elements
involved in the TVM construction. Any TCB element that directly or indirectly

> supports a TVM generation must be measured into the TVM static measurement

Reviewer 2: ```suggestion
supports a TVM  must be measured into the TVM static measurement
```

Author: Applied as well, thanks.

registers. Once a TVM is finalized, i.e. after the
`sbi_tee_host_finalize_tvm()` TH-ABI is called, the TVM static measurements
must no longer be extended.

Static measurements are stored in dedicated measurement registers and an AP-TEE
implementation must provide at least 1 and at most 8 of them. The static
measurement registers cover platform, TSM & TVM specific TCB element
measurements. The platform/TSM measurement registers may be stored in the

> hardware RoT for Measurement with platform bound life cycle. The TVM

Reviewer 2: what do you imply by platform-bound life cycle?

Author: Yes, that's what I meant. I tried to rephrase, let me know if it's clearer now.

be modified. However, a TVM guest may want to measure some of its software
components (Guest firmware, kernel OS, additional configuration parameters, etc)
while it is booting or running.

The TSM optionally provides an interface for TVM guests to extend their
measurement into dynamic measurement registers. When supporting that feature,
the TSM must store dynamic measurements separately from the TVM static ones. The
TSM can use up to 8 dynamic measurement registers for that purpose.

The TVM measurement extension interface is exposed through the optional TG-ABI
`sbi_tee_guest_extend_measurement()` FID.

=== Attestation

All above described TCB elements measurements are added to an attestation
evidence and then reported to relying parties. The attestation mechanism
and protocol that take place between the attester (i.e. the TVM) and the
remote attestation service are out of this document scope.

In this section we describe the high level attestation model for AP-TEE,
together with the attestation evidence content, format and generation process.

==== Model

The AP-TEE attestation model follows the the IETF RATS <<R1>> Remote Attestation

> architecture.

Author: Good suggestion, the PR is mentioning this now.

++++

A fixed length TCB layer identifier called `CDI_ID` can be derived from its
attestation key:

:stem: asciimath
[.center]
[stem]
++++
UDS\_ID = KeyDerivationFunction(ID_{Len}, UDS\_PubKey)
++++
:stem: asciimath
[.center]
[stem]
++++
CDI\_ID_{N} = KeyDerivationFunction(ID_{Len}, CDI\_PubKey_{N})
++++

When loading the next TCB layer, the attesting layer provides it with both its
CDI and its attestation certificate.

The TVM guest can obtain an Attestation Evidence from the TSM by calling the
TG-ABI `sbi_tee_guest_get_evidence()` FID. The guest provides a challenge value
to the TSM through this call, and that value must be included in the generated
Evidence. This value allows relying parties to establish the Attestation
Evidence freshness.

The Evidence is signed by the TSM attestation key, which is HW RoT-rooted.
A Relying Party can then verify and authenticate the Evidence with platform
owner or manufacturer provided Endorsements, like e.g. a Trust Anchor.

The TCB extension and evidence collection for a TVM attestation is shown
below:

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "Layered Attestation architecture for TVMs"]

> image::layered_attestation.svg[]

Author: I extended the `Layered Evidence` section to take this into account. Hopefully this looks better now.

It is expected that an implementation will provide implementation-specific
intrinsics to record measurements of the TSM into the firmware RoT for
measurement to support the AP-TEE layered RTMs attestation of AP-TEE workloads.

==== Evidence Format

The AP-TEE Attestation Evidence uses the IETF Entity Attestation Token
(<<EAT>>), formatted as an untagged Concise Binary Object Representation

> (<<CBOR>>) Web Token (<<CWT>>).  An AP-TEE EAT profile is proposed to narrow

Reviewer: From what I understand, the top level EAT is not a signed CWT.  Instead, it looks like an [UCCS](https://datatracker.ietf.org/doc/draft-ietf-rats-uccs/).

Author: Indeed. Thanks for spotting that one.

the EAT specification for the AP-TEE use case to enable interoperability.

> The CWT is composed of one EAT submodule Claims-Set map where the map values

Reviewer: ditto, this is not a CWT, it's an UCCS.

Author: Fixed, thanks.

> are nested tokens for the TVM, TSM and Platform Claims.

Reviewer: I'd remove _"nested"_ as it may create some confusion: the three blobs are in fact well separated into their respective slots :-)

Author: Done, thanks.

> The TVM EAT is a CWT tagged CBOR formatted token, wrapped with a

Reviewer: curiosity: why is the CWT tagging needed?

COSE_Sign1 <<COSE>> envelope. It is signed by the TSM attestation key and must
contain a TVM provided challenge, e.g. a Relying Party provided nonce for
establishing Evidence freshness.

The TSM EAT is a CWT tagged CBOR formatted token, wrapped with a
COSE_Sign1 <<COSE>> envelope. It is signed by the Platform attestation key and

> must contain a hash of the TVM public key, in order to create a cryptographic

Reviewer: I am not sure this is needed, or even possible.  IIUC, the TSM token describes the TSM layer and is created by the Platform which I don't think has any knowledge of the TVM (it'd be a layering violation), in particular about TVM's pk.

The cryptographic chaining is obtained _naturally_ using the DICE delegation logics:
```
       .---------------------------.
       | Platform measurements     |
       | Platform Attestation PK   o--.
       +---------------------------+  |
SC o-->| HRoT Att SK signature     |  |
       '---------------------------'  |
                                      |
       .---------------------------.  |
       | TSM measurements          |  |
    .--o TSM Attestation PK        |  |
    |  +---------------------------+  |
    |  | Platform Att SK signature |<-'
    |  '---------------------------'
    |  
    |  .---------------------------.
    |  | nonce                     |
    |  | TVM measurements          |
    |  +---------------------------+
    '->| TSM Att SK signature      |
       '---------------------------'
```

which has the advantage of making the life of the verifier easier :-)

Author: Yes, you are right. It does not have (and probably would be challenging implementation wise...) to be that complex. It's fixed now.

[title= "AP-TEE Attestation Evidence Certificate",scalewidth=75%]
image::attestation_certificate.svg[]

===== CBOR Attestation Certificate

CBOR formatted AP-TEE Attestation certificates are COSE_Sign1 protected CWTs,
signed with the TSM attestation key. The algorithm used to sign the certificate
is described by the COSE_Sign1 envelope. It is recommended to use an EdDSA
scheme with SHA-512, e.g. Ed25519.

The CBOR certificate COSE_Sign1 payload is a CWT which claim set is composed of
the AP-TEE evidence token and 2 additional claims:

* *Issuer* is the certificate issuer representation, i.e. the TSM `CDI_ID`.
* *Subject* is the certificate subject representation, i.e. the TVM `CDI_ID`.

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "AP-TEE CBOR Attestation Certificate"]
[source,cddl]
----
; COSE_Sign1 wrapped certificate payload
riscv-ap-tee-certificate = #6.18(riscv-ap-tee-certificate-payload)

; The payload is a CWT
riscv-ap-tee-certificate-payload = #6.61(riscv-ap-tee-claims)

issuer-label = 1 ; CWT iss
subject-label = 2 ; CWT sub
evidence-label = TBD

riscv-ap-tee-claims = (
   issuer-label => text                 ; Certificate issuer

>    subject-label => test                ; Certificate subject

Reviewer: typo:
```suggestion
   subject-label => text                ; Certificate subject
```

Author: Fixed, thanks!

   * IOMMU and translation agents
a| * All M-mode firmwares, except for the TSM-driver

 | TSM
 | N/A
a| * TEE Security Manager (TSM) and its U-mode components
   * TSM-Driver that hosts a TEEI (with TH-ABI and TG-ABI security routines). 

 | TVM
a| * Directly assigned, TEE-IO compliant devices
 | * All TVM measured pages
|===

==== TVM Measurements

At a high level, a TVM measurement is separated into a set of static
measurements and a dynamic one.

===== Static Measurements

TVM static measurements are generated from the AP-TEE workload TCB elements
involved in the TVM construction. Any TCB element that directly or indirectly
supports a TVM generation must be measured into the TVM static measurement
registers. Once a TVM is finalized, i.e. after the
`sbi_tee_host_finalize_tvm()` TH-ABI is called, the TVM static measurements
must no longer be extended.

Static measurements are stored in dedicated measurement registers and an AP-TEE
implementation must provide at least 1 and at most 16 of them. Some of the
static measurement registers cover platform or TSM specific TCB elements while
other are measuring TVM specific TCB pieces, and have respectively platform

> bound and TVM bound lifecycles. As such the former may be stored in the RoT

Reviewer 3: Sounds bit ambiguous to me. As per my understanding, static measurements cover platform, tsm or tvm measurements. Correct ?

What exactly do you mean by platform
bound and TVM bound lifecycles ? Everything static measurement is platform bound. Isn't it ?

TVM measurements are TVM bound and should be stored in TSM confidential memory.

Author: Good suggestion, thanks. Let me know if the rewording looks better to you.

for Storage while the latter should be maintained and temporarily stored in TSM
confidential memory, effectively making the TSM an extended root of trust of
measurements for the TVM.

> Which and how TCB elements are measured and stored into the static measurement

Reviewer 3: nit suggestion: The static TCB measurement mechanism and storage is implementation specific.

Author: Fixed, thanks.

registers is implementation specific. One possible layout, based on a 6 static
measurement registers AP-TEE implementation may look like the following table:

.Static Measurements Layout Example (6 registers)
[cols="^1,<3,<5,<1"]
|===
|Register |Usage |TCB Elements |Storage
 | 0
 | Platform Firmware
a| * RoT firmware
   * All SoC subsystem firmwares
   * M-mode firmware
 | RoT

 | 1
 | Platform Configuration
 | All firmware manifests
 | RoT

 | 2
 | TSM Software
a| * TSM-Driver
   * TSM
 | RoT

 | 3
 | TSM Configuration
a| * TSM-Driver manifest
   * TSM manifest
 | RoT

 | 4
 | TVM Code
 | TVM measured pages
 | TSM

 | 5
 | TVM Configuration

>  | TVM EPC and initial arguments

Reviewer 3: No earlier reference for this acronym EPC.

Author: Fixed now.

 | TSM
|===

===== Dynamic Measurements

After the TVM is finalized, static measurements are immutable and can no longer
be modified. However, a TVM guest may want to measure some of its software
components (Guest firmware, kernel OS) while it is booting.

The TSM optionally provides an interface for TVM guests to extend their

> measurement into dynamic measurement registers. When supporting that feature,

Reviewer 3: Is there a upper limit for dynamic measurement registers ?
